assembler.rb

```diff
@@ -45,6 +46,18 @@ def add_predefined_symbols
   end
 
   def assemble
+    while @parser.has_more_commands? 
+      @parser.advance
+      if @parser.command_type == :L_COMMAND
+        @symbol_table.add_entry(@parser.symbol, @instruction_address)
+      else 
+        @instruction_address += 1
+      end
+    end
+
+    @input_file.rewind
```

Comment:
Note that, as the [documentation](https://rubydoc.info/stdlib/core/IO:rewind) mentions, some `IO` objects can’t be rewound because they throw away their data once it’s been read (e.g. a `TCPSocket`). But this is only of academic interest since we know we’re using a `File` here, so it’s okay, and if in future 👀 we switch to `ARGF`, that [implements `#rewind`](https://rubydoc.info/stdlib/core/ARGF:rewind) too.